src/nwb_benchmarks/core/_streaming.py

```diff
@@ -0,0 +1,96 @@
+
+
+def read_hdf5_nwbfile_fsspec_no_cache(
+    s3_url: str,
+) -> Tuple[pynwb.NWBFile, pynwb.NWBHDF5IO, h5py.File, HTTPFile]:
+    """Read an HDF5 NWB file from an S3 URL using fsspec without a cache."""
+    (file, byte_stream) = read_hdf5_fsspec_no_cache(s3_url=s3_url)
+    io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
+    nwbfile = io.read()
+    return (nwbfile, io, file, byte_stream)
+
+
+def read_hdf5_remfile(s3_url: str) -> Tuple[h5py.File, remfile.File]:
+    """Load the raw HDF5 file from an S3 URL using remfile; does not formally read the NWB file."""
+    byte_stream = remfile.File(url=s3_url)
+    file = h5py.File(name=byte_stream)
+    return (file, byte_stream)
+
+
+def read_hdf5_nwbfile_remfile(s3_url: str) -> Tuple[pynwb.NWBFile, pynwb.NWBHDF5IO, h5py.File, remfile.File]:
+    """Read an HDF5 NWB file from an S3 URL using the ROS3 driver from h5py."""
+    (file, byte_stream) = read_hdf5_remfile(s3_url=s3_url)
+    io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
+    nwbfile = io.read()
+    return (nwbfile, io, file, byte_stream)
+
+
+def robust_ros3_read(
+    command: Callable,
+    max_retries: int = 20,
+    command_args: Union[list, None] = None,
+    command_kwargs: Union[dict, None] = None,
+) -> Any:
+    """
+    Attempt the command (usually acting on an S3 IO) up to the number of max_retries using exponential backoff.
+
+    Usually a good idea to use this to wrap any operations performed using the ROS3 driver.
+    """
+    command_args = command_args or []
+    command_kwargs = command_kwargs or dict()
+    for retry in range(max_retries):
+        try:
+            result = command(*command_args, **command_kwargs)
+            return result
+        except Exception as exception:
+            #  'cannot curl request' can show up in potentially many different error types
+            if "curl" not in str(exception):
+                raise exception
+            time.sleep(0.1 * 2**retry)
+
+    raise TimeoutError(f"Unable to complete the command ({command.__name__}) after {max_retries} attempts!")
+
+
+def read_hdf5_ros3(s3_url: str) -> h5py.File:
+    """Load the raw HDF5 file from an S3 URL using ROS3 driver; does not formally read the NWB file."""
+    ros3_form = s3_url.replace("https://dandiarchive.s3.amazonaws.com", "s3://dandiarchive")
+    aws_region = bytes("us-east-2", "ascii")  # TODO: generalize this as an argument if necessary
+    file = robust_ros3_read(
+        command=h5py.File, command_kwargs=dict(name=ros3_form, driver="ros3", aws_region=aws_region)
+    )
+    return file
+
+
+def read_hdf5_nwbfile_ros3(s3_url: str) -> Tuple[pynwb.NWBFile, pynwb.NWBHDF5IO]:
+    """Read an HDF5 NWB file from an S3 URL using the ROS3 driver from h5py."""
+    ros3_form = s3_url.replace("https://dandiarchive.s3.amazonaws.com", "s3://dandiarchive")
+    io = pynwb.NWBHDF5IO(path=ros3_form, mode="r", load_namespaces=True, driver="ros3")
+    nwbfile = robust_ros3_read(command=io.read)
```

Comment:
Does `robust_ros3_read(command=io.read)` mean that all of `io.read` will be re-executed if a failure occurs or is it smart enough to retry individual read requests from `ros3`? I assume it is the former, but i just wanted to confirm.

Reply:
If this retries all of `io.read` I'm not sure if we want to do this in the benchmarks, or instead record that a run failed, since otherwise we would sometimes record the time for a single try and at sometimes for 20 tries. At the very least, I think we would want to record the number of tries.

Reply:
Ohhh uhh

It also just occurred to me that both `remfile` and `fsspec` inherently perform retries. And I don't know if that number of retires can be tracked directly using those

Reply:
> It also just occurred to me that both `remfile` and `fsspec` inherently perform retries. And I don't know if their number of retries can be tracked directly using those

I think the difference here is though that those are retries of individual http requests, whereas `robust_ros3_read` is rerunning the entire read if one request fails. I think retries of individual connections are good and they should show up in the variation of the number of packets/data that is being transferred.

Reply:
> for that corner of the benchmarks

Do you mean just for ros3?

> , I think the best approach ... would be to add `retries` to the network tracking results before returning them.

Just to clarify, are suggesting to add a `max_retries` parameter to the `network_activity_tracker` context, such that it retries running the code that is withing the context several times before declaring a failure, e.g., ``with network_activity_tracker(tshark_path=TSHARK_PATH, max_retries=20) as network_tracker:``? That would be neat, but as far as I understand context managers, I don't think this is possible. 

One possible workaround would be to create a decorator `@retry_network_benchmark(retries=20)`. The decorator would then simply call the decorated function in a loop and record: 1) the number of retries,  2) the results of the final run, and 3) the total time it took to complete the operation. 

```
@retry_network_benchmark(retries=20)
def track_network_activity_during_read(self, s3_url: str):
        with network_activity_tracker(tshark_path=TSHARK_PATH) as network_tracker:
            self.file, self.bytestream = read_hdf5_fsspec_no_cache(s3_url=s3_url)
        return network_tracker.asv_network_statistics
```

If we also want to track the network traffic failed tries, then I think we would need to move the network tracking and result generation completely into the decorator, i.e., a test would then look something like:

```
@network_benchmark(retries=20)
def track_network_activity_during_read(self, s3_url: str):
       self.file, self.bytestream = read_hdf5_fsspec_no_cache(s3_url=s3_url)
```

Reply:
> Do you mean just for ros3?

Just for ros3, in the custom network tracking tests - the number of retries can be reported along with the packet info

I added this in last night - all it does it report back the number of enumerations from within the `robust_ros3` call if the `retry` is enabled

Exposure in `robust_ros3` call

https://github.com/NeurodataWithoutBorders/nwb_benchmarks/blob/refactor_functional_fully_explicit/src/nwb_benchmarks/core/_streaming.py#L75-L78

Passing from associated file readers

https://github.com/NeurodataWithoutBorders/nwb_benchmarks/blob/refactor_functional_fully_explicit/src/nwb_benchmarks/core/_streaming.py#L101-L108

https://github.com/NeurodataWithoutBorders/nwb_benchmarks/blob/refactor_functional_fully_explicit/src/nwb_benchmarks/core/_streaming.py#L128-L133

Then add to values tracked by the context, but not as apart of the context itself

https://github.com/NeurodataWithoutBorders/nwb_benchmarks/blob/refactor_functional_fully_explicit/src/nwb_benchmarks/benchmarks/network_tracking_remote_slicing.py#L69-L72

Reply:
Looks good to me. @sinha-r  this is something you should be aware of when implementing tests with `ros3` . The issue is that `ros3` does not itself retry individual network requests, so it is more likely to fail, since reading an NWB files results in many requests. The approach that @CodyCBakerPhD added repeats the entire NWB read several times in case it fails. `fsspec` and `remfile` retry connections themselves so we don't need to worry about it there.